### PR TITLE
Have dependency version policy

### PR DESCRIPTION
**Pull request by brenthuisman**
As per #1731, this PR adds a dependencies version policy to the docs. It also includes a file containing a list of all configure-time deps (please check!).

Obsoletes https://github.com/arbor-sim/arbor/wiki/Testing-and-version-management

Discussion point: We should add maximum versions, or at least take care that we don't outrun environments where we need to have Arbor. On Ebrains, our most recent `pybind` bump turned out to be incompatible with another projects maximum version. The user platforms implicitly impose maximum versions, but maybe people will want to define this in another way.

### CODE REVIEW

File: doc/contrib/dependency-management.rst [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+2. Dependencies shall be set to a (commit hash corresponding to a) specific version tag. (All current dependencies use semver.)
+3. Git submodules and dependencies in ``spack/package.py`` shall be in sync.
+4. The version shall be compatible with the user platforms (see above).
+5. The version shall be compatible with the requirements in ``doc/dependencies.csv``.
+6. Submodules shall not be updated past the most recent version of the dependency in Spack.
+
+   * This prevents Spack builds from pulling in ``master``, when a more recent version than available is required. `See here <https://spack.readthedocs.io/en/latest/packaging_guide.html#version-comparison>`_.
+   * This is a manual check, e.g. for pybind: `see pybind package.py <https://github.com/spack/spack/blob/develop/var/spack/repos/builtin/packages/py-pybind11/package.py>`_
+7. Actually updating shall remain a manual process. Update may require nontrivial updates to Arbor, and updates to Spack upstream (e.g. make PR for pybind update).
+8. A dependency update shall have a separate PR, and such PR updates a single dependency at a time.
```
Comment (thorstenhater) [inline doc/contrib/dependency-management.rst:53]: `and such *a* PR updates a`

But that might not work:
- Dependencies _might_ introduce changes to the code in Arbor
- Also, somewhat theoretical, a dependency might require updating another.

File: doc/install/build_install.rst [hunk 2]
```diff
@@ -22,34 +22,18 @@ Minimum requirements
 The non distributed (i.e. no MPI) version of Arbor can be compiled on Linux or OS X systems
 with very few tools.
```
Comment (thorstenhater) [inline doc/install/build_install.rst:24]: I suggest adding small snippets on how to build Arbor on these platforms
* Ubuntu LTS
* MacOS
* Piz Daint
* Juwels Booster

picked from above.

Comment (Helveg) [reply doc/install/build_install.rst:24]: Yes please, a build guide catered to  the supported platforms would be incredibly helpful.

Comment (brenthuisman) [reply doc/install/build_install.rst:24]: You want this in this PR? The `build_install.rst` could use some more cleaning up IMHO, but I didn't want to merge that sort of change in this PR.

**Comment by brenthuisman** [comment]
Summary of team meeting on this subject today:

* submodularize (to be done in other PR)
    * ext/json
    * ext/random123
    * test/gtest.h
        * also move to ext/
* move pybind to ext/ (to be done in other PR)
* in `dependencies.csv` in #1865
    * [x] distinguish tool from dependency
    * [ ] add above submodules
* bump versions policy:
    * [x] leave submodules to lowest possible
    * [ ] modify ``.github/workflows/check-submodules.yml`` to actually pull and test.

File: doc/contrib/dependency-management.rst [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+* Ubuntu LTS-latest
+* Ubuntu LTS-latest-1
+* MacOS-latest
+* MacOS-latest-1
+* Cray programming environment on Piz Daint
+* Programming environment on Juwels Booster (**todo** CI at JSC)
+* Github Action venvs, see `list <https://github.com/actions/virtual-environments>`_.
+* Manylinux containers. For compatibility of manylinux tags, see `here <https://github.com/pypa/manylinux#readme>`_.
+
+Dependency update rules
```
[the remark this answers is not in the document]
Comment (brenthuisman) [reply doc/contrib/dependency-management.rst:45]: We chose to not do that earlier, see rest of docs.

**Comment by brenthuisman** [comment]
``doc/dependencies.csv`` is complete. Not entirely happy with the way it is presented on the build_install page: I want to avoid duplicating info, but having people lookup things in the table isn't nice. The `csv-table` does not let me hide column and rows. I've looked into a few plugins [1],[2] to give that flexiblity but they don't parse any `reST`, so the lists in the `csv` would become hard to understand.

Another option is to split out the compilers into their own csv, and lose the single-file-has-source-of-truth idea.

[1] https://github.com/sphinx-contrib/datatemplates
[2] https://pythonhosted.org/sphinxcontrib-programoutput/

File: doc/install/build_install.rst (hunk 2 above)
Comment (thorstenhater) [reply doc/install/build_install.rst:51]: Fine with adding it there.

File: doc/contrib/dependency-management.rst [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+
+.. csv-table:: List of configure time dependencies
+   :file: ../dependencies.csv
+   :widths: 10, 20, 10, 70, 1
+   :header-rows: 1
+
+User platforms
+--------------
+
+Although Arbor should in principle run everywhere modern compilers are found, a couple of user platforms impose hard requirements on Arbor compatiblity. That is: Arbor should build without hassle on this set of platforms. If not, we have a bug.
```
[the remark this answers is not in the document]
Comment (thorstenhater) [reply doc/contrib/dependency-management.rst]: Did you change anything here?

**Review by thorstenhater — CHANGES_REQUESTED** [review]

**Review by thorstenhater — APPROVED** [review]

**Pull request merged**